File: pypy/benchmark_mono.py
"""Benchmark polysum with a MONOMORPHIC call site, for comparison with
benchmark_poly.py.

Identical loop and element count, but every summand is the same type
(AddInt), so the `s.add(total)` call site sees exactly one target. The JIT
can inline it behind a single class guard that never fails -- isolating the
cost that the random 3-type mix adds in benchmark_poly.py.

    python3 benchmark_mono.py
    pypy3 benchmark_mono.py
"""

import platform
import random
import time

from summands import AddInt, polysum

N = 5_000_000  # number of summands (matches benchmark_poly.py)
REPEATS = 5  # timed repetitions; we report the best one
SEED = 1234


def build_data(n):
    rng = random.Random(SEED)
    # Single type, random small int value for each summand.
    return [AddInt(rng.randint(0, 9)) for _ in range(n)]


def main():
    data = build_data(N)

    # Warm-up: lets PyPy's JIT compile the loop before we start timing.
    polysum(data)

    best = float("inf")
    result = None
    for _ in range(REPEATS):
        start = time.perf_counter()
        result = polysum(data)
        elapsed = time.perf_counter() - start
        best = min(best, elapsed)

    impl = platform.python_implementation()
    version = platform.python_version()
    rate = N / best

    print(f"{impl} {version}")
    print(f"  summands    : {N:,} (monomorphic: all AddInt)")
    print(f"  repeats     : {REPEATS}")
    print(f"  best time   : {best:.4f} s")
    print(f"  throughput  : {rate/1e6:.1f} M calls/s")
    print(f"  (sum check) : {result}")


if __name__ == "__main__":
    main()

File: pypy/benchmark_mono.js
"use strict";
// JS port of benchmark_mono.py: polysum over 5M summands, all one type
// (monomorphic call site), for comparison with benchmark_poly.js.
const { AddInt, polysum } = require("./summands");
const { makeRng, timeBest, report } = require("./bench_util");

const N = 5_000_000;
const REPEATS = 5;
const SEED = 1234;

function buildData(n) {
  const rng = makeRng(SEED);
  const data = new Array(n);
  for (let i = 0; i < n; i++) data[i] = new AddInt(Math.floor(rng() * 10));
  return data;
}

const data = buildData(N);
polysum(data); // warm-up
const { best, result } = timeBest(() => polysum(data), REPEATS);
report({
  label: "polysum, monomorphic (all AddInt)",
  n: N,
  unit: "calls/s",
  best,
  result,
});
